core/drivers/crypto/hisilicon/hpre_dh.c

```diff
@@ -0,0 +1,380 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2022, HiSilicon Technologies Co., Ltd.
+ * Kunpeng hardware accelerator hpre dh algorithm implementation.
+ */
+
+#include <hpre_main.h>
+#include <hpre_dh.h>
+
+static TEE_Result hpre_dh_alloc_keypair(struct dh_keypair *key,
+					size_t size_bits)
+{
+	if (!key || !size_bits) {
+		EMSG("Invalid input parameter!");
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	memset(key, 0, sizeof(*key));
+	key->g = crypto_bignum_allocate(size_bits);
+	if (!key->g)
+		goto g_err;
+
+	key->p = crypto_bignum_allocate(size_bits);
+	if (!key->p)
+		goto p_err;
+
+	key->x = crypto_bignum_allocate(size_bits);
+	if (!key->x)
+		goto x_err;
+
+	key->y = crypto_bignum_allocate(size_bits);
+	if (!key->y)
+		goto y_err;
+
+	/* Allocate subprime even if not used */
+	key->q = crypto_bignum_allocate(size_bits);
+	if (!key->q)
+		goto q_err;
+
+	return TEE_SUCCESS;
+q_err:
+	crypto_bignum_free(&key->y);
+y_err:
+	crypto_bignum_free(&key->x);
+x_err:
+	crypto_bignum_free(&key->p);
+p_err:
+	crypto_bignum_free(&key->g);
+g_err:
+	EMSG("HPRE dh alloc key pair fail.");
+
+	return TEE_ERROR_OUT_OF_MEMORY;
+}
+
+static enum hisi_drv_status hpre_dh_fill_sqe(void *bd, void *info)
+{
+	struct hpre_dh_msg *msg = info;
+	struct hpre_sqe *sqe = bd;
+
+	sqe->w0 = msg->alg_type | (0x1 << HPRE_DONE_SHIFT);
```

Comment:
Please use BIT(HPRE_DONE_SHIFT) or SHIFT_U32(0x1, HPRE_DONE_SHIFT) depending on whether it's a single bit or a field.

Reply:
solved